tests/rados/pool_tests.py

```diff
@@ -400,3 +400,43 @@ def run(ceph_cluster, **kw):
             log.info(f"Completed the test of pg_min_num on pool: {entry['pool_name']} ")
         log.info("pg_min_num tests completed")
         return 0
+
+    if config.get("verify_pg_num_limit"):
+        """
+        This test is to verify the pg number limit for a pools.
+        Script cover the following steps-
+           1. Creating a pool with default pg number
+           2. Set the pg number more than the limit that is more than 128
+           3. Cheking that pg number getting reduced or not
+        """
+
+        pool_name = config.get("verify_pg_num_limit")["pool_name"]
+        rados_obj.create_pool(pool_name)
+
+        # set the pg number more than 128
+        cmd = f"ceph osd pool set {pool_name} pg_num 256"
+        rados_obj.run_ceph_command(cmd=cmd)
+        time.sleep(5)
+
+        # Checking the pg number is getting below 128
+        pg_num_get = f"ceph osd pool get {pool_name} pg_num"
+        time_end = time.time() + 60 * 30
```

Comment:
Can we have some data in the pool, so that PG merge happens while moving data?

Reply:
I noticed that creating objects,  pg_number reduction is taking more time.
For the following objects in the test_pool it took around 3 days. That is the reason I have not added the data.

[ceph: root@ceph-test-rados-lu54di-node1-installer /]# ceph df
--- RAW STORAGE ---
CLASS     SIZE    AVAIL    USED  RAW USED  %RAW USED
hdd    275 GiB  228 GiB  47 GiB    47 GiB      17.07
TOTAL  275 GiB  228 GiB  47 GiB    47 GiB      17.07
 
--- POOLS ---
POOL                   ID  PGS   STORED  OBJECTS     USED  %USED  MAX AVAIL
device_health_metrics   1    1   13 KiB       15   39 KiB      0     90 GiB
cephfs.cephfs.meta      2   16   53 KiB       22  265 KiB      0     91 GiB
cephfs.cephfs.data      3   32      0 B        0      0 B      0     90 GiB
.rgw.root               4   32  2.2 KiB        4   60 KiB      0     98 GiB
default.rgw.log         5   32  4.3 KiB      209  484 KiB      0     91 GiB
default.rgw.control     6   32      0 B        8      0 B      0     90 GiB
default.rgw.meta        7    8    509 B        2   32 KiB      0     90 GiB
test_pool               8   64   18 GiB    1.97M   54 GiB  16.74     91 GiB
[ceph: root@ceph-test-rados-lu54di-node1-installer /]#

Reply:
Please suggest if any changes are required?

Reply:
Couple of things that can be done:
1.  Increase the backfill and recovery threads, so that data is written faster.
2.  Write Large objects in small numbers. Small objects take longer to move for the same amount of pool filled. Also, less data can be written for simulation.
3.  You can start deleting some objects, so that we will have less objects as time progresses, which would simulate hybrid workloads on the pools, also fastening the process.

Let me know if this helps.

Reply:
Also, as you confirmed above that PG_autoscaler should be on for this, the target PG num for any pool created can be checked in autoscale status. The current PG num and New PG num is reflected. This would give assurance that the cluster is in process of changing the PG count.

``` # ceph osd pool autoscale-status
POOL                         SIZE  TARGET SIZE  RATE  RAW CAPACITY   RATIO  TARGET RATIO  EFFECTIVE RATIO  BIAS  PG_NUM  NEW PG_NUM  AUTOSCALE  BULK
test                           0                 4.0        249.9G  0.0000                                  1.0     256          32  on         False
 ```